### PR TITLE
Allow for user defined image size

### PR DESCRIPTION
**Pull request by jkozol**
When starting a compose users can pass in an image size. The resulting image will then be of this size. If no image size is passed a default size will be used.

### CODE REVIEW

File: internal/distro/fedora30/distro.go [hunk 1]
```diff
@@ -129,7 +135,8 @@ func New(confPaths []string) *Fedora30 {
 		},
 		KernelOptions: "ro biosdevname=0 net.ifnames=0",
 		Bootable:      false,
-		Assembler:     func(uefi bool) *pipeline.Assembler { return r.rawFSAssembler("filesystem.img") },
+		DefaultSize:   3221225472,
```
Comment (teg) [inline internal/distro/fedora30/distro.go]: `3 * GigaByte` ?

File: internal/distro/rhel82/distro.go [hunk 2]
```diff
@@ -187,7 +189,8 @@ func New(confPaths []string) *RHEL82 {
 		},
 		Bootable:      false,
 		KernelOptions: "ro net.ifnames=0",
-		Assembler:     func(uefi bool) *pipeline.Assembler { return r.rawFSAssembler("filesystem.img") },
+		DefaultSize:   3221225472,
```
Comment (teg) [inline internal/distro/rhel82/distro.go]: `3 * GigaByte` ?

File: internal/distro/fedora30/distro.go (hunk 1 above)
Comment (teg) [reply internal/distro/fedora30/distro.go]: Don't know how to comment on the commit message, so leaving this here: please flesh it out to make it clear what you are adding (allowing size to be passed in) and changing (per output-type default sizes (as opposed to?), for when no size is passed in).

File: cmd/osbuild-pipeline/main.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ func main() {
 		panic(err.Error())
 	}
 
-	pipeline, err := d.Pipeline(blueprint, nil, checksums, archArg, format)
+	pipeline, err := d.Pipeline(blueprint, nil, checksums, archArg, format, 0)
```
Comment (teg) [inline cmd/osbuild-pipeline/main.go]: This needs to be part of the previous commit, or that will not compile, right?

File: internal/pipeline/tar_assembler.go [hunk 4]
```diff
@@ -6,16 +6,18 @@ package pipeline
...
+	Size        uint64 `json:"size"`
 	Compression string `json:"compression,omitempty"`
 }
 
 func (TarAssemblerOptions) isAssemblerOptions() {}
 
 // NewTarAssemblerOptions creates a new TarAssemblerOptions object, with the
 // mandatory options set.
-func NewTarAssemblerOptions(filename string) *TarAssemblerOptions {
+func NewTarAssemblerOptions(filename string, size uint64) *TarAssemblerOptions {
```
Comment (teg) [inline internal/pipeline/tar_assembler.go]: Hm, this won't compile, will it? You need to update the callers at the same time (ideally without any functional change).

File: internal/store/store.go [hunk 5]
```diff
@@ -463,7 +463,7 @@ func (s *Store) PushCompose(composeID uuid.UUID, bp *blueprint.Blueprint, checks
 		repos = append(repos, source.RepoConfig())
 	}
 
-	pipeline, err := s.distro.Pipeline(bp, repos, checksums, arch, composeType)
+	pipeline, err := s.distro.Pipeline(bp, repos, checksums, arch, composeType, size)
```
Comment (teg) [inline internal/store/store.go]: I think you should change this to pass 0 in the commit that changes the interface, then update in this commit (introducing the size parameter to PushCompose()) to pass on the real size.

File: internal/store/store.go [hunk 6]
```diff
@@ -443,7 +443,7 @@ func (s *Store) GetComposeResult(id uuid.UUID) (io.ReadCloser, error) {
 	return os.Open(*s.stateDir + "/outputs/" + id.String() + "/result.json")
 }
 
-func (s *Store) PushCompose(composeID uuid.UUID, bp *blueprint.Blueprint, checksums map[string]string, arch, composeType string, uploadTarget *target.Target) error {
+func (s *Store) PushCompose(composeID uuid.UUID, bp *blueprint.Blueprint, checksums map[string]string, arch, composeType string, size uint64, uploadTarget *target.Target) error {
```
Comment (teg) [inline internal/store/store.go]: You need to update the callers at the same time (ideally just passing in zero), to make sure things compile.

File: internal/weldr/api.go [hunk 7]
```diff
@@ -1246,6 +1248,11 @@ func (api *API) composeHandler(writer http.ResponseWriter, request *http.Request
 
 	bp := api.store.GetBlueprintCommitted(cr.BlueprintName)
 
+	size := cr.Size
+	if size%MegaByte != 0 {
+		size = (size/MegaByte + 1) * MegaByte
```
Comment (teg) [inline internal/weldr/api.go]: This needs some discussion / comment. The alternative is to refuse the compose if we don't like what is passed (we do that further down the stack). I think you are right that this is what we want, but would need some discussion as to why at this point?

File: test/cases/ami_empty_blueprint.json [hunk 8]
```diff
@@ -130,7 +130,7 @@
       "options": {
         "format": "raw.xz",
         "filename": "image.raw.xz",
-        "size": 3222274048,
+        "size": 6442450944,
```
Comment (teg) [inline test/cases/ami_empty_blueprint.json]: Why are the sizes changing? Should be some argument why the new sizes are right wherever the change was introduced. Also, this needs to be squashed with the commit that introduce the change to make sure the tests keep passing.

File: internal/distro/fedora30/distro.go (hunk 1 above)
Comment (larskarlitski) [reply internal/distro/fedora30/distro.go]: > `3 * GigaByte` ?

This seems to be a go-ism. I've seen this done for time and kind of liked the idea: https://gobyexample.com/timeouts

**Pull request merged**